apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1217 @@
+#include "../../model/world/columnbase.hpp"
+#include "../../model/world/commandmacro.hpp"
+#include "../../model/world/commands.hpp"
+#include "../../model/world/data.hpp"
+#include "../../model/world/idtable.hpp"
+#include "../../model/world/idtree.hpp"
+#include "../../model/world/land.hpp"
+#include "../../model/world/resourcetable.hpp"
+#include "../../model/world/tablemimedata.hpp"
+#include "../../model/world/universalid.hpp"
+
+#include "editmode.hpp"
+#include "pagedworldspacewidget.hpp"
+#include "mask.hpp"
+#include "object.hpp" // Something small needed regarding pointers from here ()
+#include "terrainselection.hpp"
+#include "worldspacewidget.hpp"
+
+CSVRender::TerrainShapeMode::TerrainShapeMode (WorldspaceWidget *worldspaceWidget, osg::Group* parentNode, QWidget *parent)
+: EditMode (worldspaceWidget, QIcon {":scenetoolbar/editing-terrain-shape"}, Mask_Terrain | Mask_Reference, "Terrain land editing", parent),
+    mBrushSize(0),
+    mBrushShape(0),
+    mShapeBrushScenetool(0),
+    mDragMode(InteractionType_None),
+    mParentNode(parentNode),
+    mIsEditing(false),
+    mTotalDiffY(0),
+    mShapeEditTool(0),
+    mShapeEditToolStrength(8),
+    mTargetHeight(0)
+{
+}
+
+void CSVRender::TerrainShapeMode::activate(CSVWidget::SceneToolbar* toolbar)
+{
+    if (!mTerrainShapeSelection)
+    {
+        mTerrainShapeSelection.reset(new TerrainSelection(mParentNode, &getWorldspaceWidget(), TerrainSelectionType::Shape));
+    }
+
+    if(!mShapeBrushScenetool)
+    {
+        mShapeBrushScenetool = new CSVWidget::SceneToolShapeBrush (toolbar, "scenetoolshapebrush", getWorldspaceWidget().getDocument());
+        connect(mShapeBrushScenetool, SIGNAL (clicked()), mShapeBrushScenetool, SLOT (activate()));
+        connect(mShapeBrushScenetool->mShapeBrushWindow, SIGNAL(passBrushSize(int)), this, SLOT(setBrushSize(int)));
+        connect(mShapeBrushScenetool->mShapeBrushWindow, SIGNAL(passBrushShape(int)), this, SLOT(setBrushShape(int)));
+        connect(mShapeBrushScenetool->mShapeBrushWindow->mSizeSliders->mBrushSizeSlider, SIGNAL(valueChanged(int)), this, SLOT(setBrushSize(int)));
+        connect(mShapeBrushScenetool->mShapeBrushWindow->mToolSelector, SIGNAL(currentIndexChanged(int)), this, SLOT(setShapeEditTool(int)));
+        connect(mShapeBrushScenetool->mShapeBrushWindow->mToolStrengthSlider, SIGNAL(valueChanged(int)), this, SLOT(setShapeEditToolStrength(int)));
+    }
+
+    EditMode::activate(toolbar);
+    toolbar->addTool (mShapeBrushScenetool);
+}
+
+void CSVRender::TerrainShapeMode::deactivate(CSVWidget::SceneToolbar* toolbar)
+{
+    EditMode::deactivate(toolbar);
+}
+
+void CSVRender::TerrainShapeMode::primaryOpenPressed (const WorldspaceHitResult& hit) // Apply changes here
+{
+}
+
+void CSVRender::TerrainShapeMode::primaryEditPressed(const WorldspaceHitResult& hit)
+{
+    mCellId = getWorldspaceWidget().getCellId (hit.worldPos);
+
+    if (hit.hit && hit.tag == 0)
+    {
+    }
```

Comment:
Probably this should be removed.

Reply:
Whoops. Needs a call to editTerrainShapeGrid. Clicking the scene should also have an effect.